### PR TITLE
test-bot: print output if ARGV.verbose?

### PR DESCRIPTION
**Pull request by xu-cheng**

### CODE REVIEW

**Comment by DomT4** [comment]
Thanks for running with this Xu. You did a better job than me by far on tweaking the code to allow more verbose output - Shall delete my local branch with a sigh of relief :smiley_cat: :tada:.

**Comment by xu-cheng** [comment]
En, `IO.popen` whose version I needed is not available in ruby 1.8. Any suggestions?

File: Library/Homebrew/cmd/test-bot.rb [hunk 1]
```diff
@@ -112,33 +112,37 @@ def run
...
-        File.open(log, "wb") do |f|
-          STDOUT.reopen(f)
-          STDERR.reopen(f)
-        end
-        Dir.chdir(@repository) if @command.first == "git"
-        exec(*@command)
+      verbose = ARGV.verbose?
+      success = nil
+      working_dir = @command.first == "git" ? @repository : Dir.pwd
+      logf = File.open(log, "wb")
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: Whats the `f`: `file`? If so: `log_file`

File: Library/Homebrew/cmd/test-bot.rb [hunk 2]
```diff
@@ -112,33 +112,37 @@ def run
...
+      puts if verbose
+
+      IO.popen(@command, :err => [:child, :out], :chdir => working_dir) do |pipe|
+         while buf = pipe.gets
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: `buf` => `buffer`

Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: or just `line`

File: Library/Homebrew/cmd/test-bot.rb [hunk 3]
```diff
@@ -112,33 +112,37 @@ def run
...
-        if has_output? and (failed? or @puts_output_on_success)
-          puts @output
-        end
-        FileUtils.rm(log) unless ARGV.include? "--keep-logs"
+      if has_output? && (failed? or @puts_output_on_success) && !verbose
+        puts @output
       end
 
+      logf.flush
+      logf.close
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: These should probably be nearer the last `logf` usage.

File: Library/Homebrew/cmd/test-bot.rb [hunk 4]
```diff
@@ -112,33 +112,37 @@ def run
...
+      FileUtils.rm(log) unless ARGV.include? "--keep-logs"
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: Seems weird to write a file and then delete it (although I know that's my existing code). Maybe just read the output into a string buffer and only write it to disk if we want to keep it?

**Comment by mikemcquaid** [comment]
> En, IO.popen whose version I needed is not available in ruby 1.8. Any suggestions?

Open3#popen3

**Comment by mikemcquaid** [comment]
Nice work here :+1:

File: Library/Homebrew/cmd/test-bot.rb (hunk 4 above)
Comment (xu-cheng) [inline Library/Homebrew/cmd/test-bot.rb]: > Maybe just read the output into a string buffer and only write it to disk if we want to keep it?

Yeah, I did this at first. But I think if something failed, we may have no log at all.

**Comment by xu-cheng** [comment]
> Open3#popen3

Still requires Ruby 1.9+ to redirect stderr. Unless we back port the code using [this gem](https://github.com/kindkid/open3_backport).

**Comment by mikemcquaid** [comment]
Check the Homebrew code and use the same logic we do for writing logfiles?

**Comment by xu-cheng** [comment]
> Check the Homebrew code and use the same logic we do for writing log files?

Good advice. I reimplement this based on `Formula#system`

**Comment by xu-cheng** [comment]
comments on new implementation?

File: Library/Homebrew/cmd/test-bot.rb [hunk 5]
```diff
@@ -112,33 +112,47 @@ def run
...
       end
 
+      @output = ""
       start_time = Time.now
+      verbose = ARGV.verbose?
+      working_dir = Pathname.new(@command.first == "git" ? @repository : Dir.pwd)
+      puts if verbose
 
-      log = log_file_path
+      File.open(log_file_path, "wb") do |log|
```
Comment (mikemcquaid) [inline Library/Homebrew/cmd/test-bot.rb]: I'd perhaps consider (as I mentioned before) only opening the file if there's a non-empty string to write out.

Comment (xu-cheng) [inline Library/Homebrew/cmd/test-bot.rb]: Updated

**Comment by mikemcquaid** [comment]
Looks good. Feel free to merge when you've tested and are happy with it in both modes locally.

**Pull request closed without merging**